File: src/engine/lexer.ts
/**
 * Hand-written tokenizer for the Riverbed SQL dialect.
 *
 * Deliberately not a regex soup: a single left-to-right scan that tracks source
 * offsets, so parse errors can point at the exact character the user typed. The
 * editor underlines the offending token using these offsets, which is the
 * difference between "syntax error" and a message someone can act on.
 */

export type TokType =
  | 'ident'
  | 'number'
  | 'string'
  | 'keyword'
  | 'op'
  | 'lparen'
  | 'rparen'
  | 'comma'
  | 'star'
  | 'duration'
  | 'eof';

export interface Token {
  type: TokType;
  value: string;
  start: number;
  end: number;
}

export const KEYWORDS = new Set([
  'select', 'from', 'where', 'group', 'by', 'order', 'limit', 'window',
  'and', 'or', 'not', 'as', 'asc', 'desc', 'distinct', 'like', 'in',
  'is', 'null', 'true', 'false', 'case', 'when', 'then', 'else', 'end',
]);

/** Longest-first, so '>=' is matched before '>'. */
const OPERATORS = ['>=', '<=', '!=', '<>', '||', '=', '<', '>', '+', '-', '*', '/', '%'];

const DURATION_UNITS: Record<string, number> = { s: 1, m: 60, h: 3600, d: 86400 };

/**
 * A user-facing SQL error carrying source offsets for underlining.
 *
 * Everything the user can trigger by typing goes through this class, so the UI
 * can render a structured error instead of a stack trace.
 */
export class SqlError extends Error {
  readonly start: number;
  readonly end: number;
  readonly hint: string | null;

  constructor(message: string, start = 0, end = 0, hint: string | null = null) {
    super(message);
    this.name = 'SqlError';
    this.start = start;
    this.end = end;
    this.hint = hint;
  }

  toJSON() {
    return { message: this.message, start: this.start, end: this.end, hint: this.hint };
  }
}

const isDigit = (c: string) => c >= '0' && c <= '9';
const isAlpha = (c: string) => /[A-Za-z_]/.test(c);
const isAlnum = (c: string) => /[A-Za-z0-9_]/.test(c);

export function tokenize(src: string): Token[] {
  const tokens: Token[] = [];
  const n = src.length;
  let i = 0;

  while (i < n) {
    const c = src[i]!;

    if (c === ' ' || c === '\t' || c === '\r' || c === '\n') {
      i += 1;
      continue;
    }

    // Line comments.
    if (c === '-' && src[i + 1] === '-') {
      const j = src.indexOf('\n', i);
      i = j === -1 ? n : j + 1;
      continue;
    }

    // Single-quoted strings; '' escapes an embedded quote.
    if (c === "'") {
      let j = i + 1;
      const buf: string[] = [];
      for (;;) {
        if (j >= n) {
          throw new SqlError('Unterminated string literal.', i, n, 'Add a closing single quote.');
        }
        if (src[j] === "'") {
          if (src[j + 1] === "'") {
            buf.push("'");
            j += 2;
            continue;
          }
          break;
        }
        buf.push(src[j]!);
        j += 1;
      }
      tokens.push({ type: 'string', value: buf.join(''), start: i, end: j + 1 });
      i = j + 1;
      continue;
    }

    // Numbers, and durations such as 5m / 30s / 2h / 1d.
    if (isDigit(c) || (c === '.' && isDigit(src[i + 1] ?? ''))) {
      let j = i;
      let seenDot = false;
      while (j < n && (isDigit(src[j]!) || (src[j] === '.' && !seenDot))) {
        if (src[j] === '.') seenDot = true;
        j += 1;
      }
      // A bare unit letter directly after the digits makes it a duration.
      const unit = src[j] ?? '';
      if (!seenDot && unit in DURATION_UNITS) {
        const next = src[j + 1] ?? '';
        if (next === '' || !isAlnum(next)) {
          tokens.push({ type: 'duration', value: src.slice(i, j + 1), start: i, end: j + 1 });
          i = j + 1;
          continue;
        }
      }
      tokens.push({ type: 'number', value: src.slice(i, j), start: i, end: j });
      i = j;
      continue;
    }

    // Identifiers and keywords.
    if (isAlpha(c)) {
      let j = i;
      while (j < n && isAlnum(src[j]!)) j += 1;
      const word = src.slice(i, j);
      tokens.push({
        type: KEYWORDS.has(word.toLowerCase()) ? 'keyword' : 'ident',
        value: word,
        start: i,
        end: j,
      });
      i = j;
      continue;
    }

    // Double-quoted identifiers, for names that collide with keywords.
    if (c === '"') {
      const j = src.indexOf('"', i + 1);
      if (j === -1) throw new SqlError('Unterminated quoted identifier.', i, n);
      tokens.push({ type: 'ident', value: src.slice(i + 1, j), start: i, end: j + 1 });
      i = j + 1;
      continue;
    }

    if (c === '(') { tokens.push({ type: 'lparen', value: c, start: i, end: i + 1 }); i += 1; continue; }
    if (c === ')') { tokens.push({ type: 'rparen', value: c, start: i, end: i + 1 }); i += 1; continue; }
    if (c === ',') { tokens.push({ type: 'comma', value: c, start: i, end: i + 1 }); i += 1; continue; }

    const op = OPERATORS.find((candidate) => src.startsWith(candidate, i));
    if (op) {
      // '*' gets its own type so `SELECT *` and `a * b` are distinguishable
      // without lookbehind hacks in the parser.
      tokens.push({ type: op === '*' ? 'star' : 'op', value: op, start: i, end: i + op.length });
      i += op.length;
      continue;
    }

    throw new SqlError(`Unexpected character '${c}'.`, i, i + 1);
  }

  tokens.push({ type: 'eof', value: '', start: n, end: n });
  return tokens;
}

/** '5m' -> 300. Assumes the lexer already validated the shape. */
export function parseDuration(text: string): number {
  const unit = text[text.length - 1]!;
  return Number(text.slice(0, -1)) * DURATION_UNITS[unit]!;
}

File: server/riverbed/sql/lexer.py
"""Hand-written tokenizer for the Riverbed SQL dialect.

Deliberately not a regex soup: a single left-to-right scan that tracks byte
offsets, so parse errors can point at the exact character the user typed. The
frontend underlines the offending token using these offsets, which is the
difference between "syntax error" and an error message someone can act on.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto


class TokType(Enum):
    IDENT = auto()
    NUMBER = auto()
    STRING = auto()
    KEYWORD = auto()
    OP = auto()
    LPAREN = auto()
    RPAREN = auto()
    COMMA = auto()
    STAR = auto()
    DURATION = auto()  # 30s, 5m, 2h, 1d
    EOF = auto()


KEYWORDS = {
    "select", "from", "where", "group", "by", "order", "limit", "window",
    "and", "or", "not", "as", "asc", "desc", "distinct", "like", "in",
    "is", "null", "true", "false", "case", "when", "then", "else", "end",
}

# Longest-first so that '>=' is matched before '>'.
OPERATORS = (
    ">=", "<=", "!=", "<>", "=", "<", ">", "+", "-", "*", "/", "%", "||",
)

DURATION_UNITS = {"s": 1, "m": 60, "h": 3600, "d": 86400}


@dataclass(frozen=True)
class Token:
    type: TokType
    value: str
    start: int
    end: int

    def __repr__(self) -> str:  # pragma: no cover - debugging aid
        return f"Token({self.type.name}, {self.value!r})"


class SqlError(Exception):
    """A user-facing SQL error carrying source offsets for underlining.

    Every error the user can trigger by typing goes through this class, so the
    API layer can return a structured error instead of a stack trace.
    """

    def __init__(self, message: str, start: int = 0, end: int = 0, hint: str | None = None):
        super().__init__(message)
        self.message = message
        self.start = start
        self.end = end
        self.hint = hint

    def to_dict(self) -> dict:
        return {
            "message": self.message,
            "start": self.start,
            "end": self.end,
            "hint": self.hint,
        }


def tokenize(src: str) -> list[Token]:
    tokens: list[Token] = []
    i = 0
    n = len(src)

    while i < n:
        c = src[i]

        # Whitespace
        if c in " \t\r\n":
            i += 1
            continue

        # Line comments
        if src.startswith("--", i):
            j = src.find("\n", i)
            i = n if j == -1 else j + 1
            continue

        # Single-quoted string literals, '' escapes an embedded quote.
        if c == "'":
            j = i + 1
            buf: list[str] = []
            while True:
                if j >= n:
                    raise SqlError("Unterminated string literal.", i, n,
                                   hint="Add a closing single quote.")
                if src[j] == "'":
                    if j + 1 < n and src[j + 1] == "'":
                        buf.append("'")
                        j += 2
                        continue
                    break
                buf.append(src[j])
                j += 1
            tokens.append(Token(TokType.STRING, "".join(buf), i, j + 1))
            i = j + 1
            continue

        # Numbers, and durations like 5m / 30s / 2h / 1d.
        if c.isdigit() or (c == "." and i + 1 < n and src[i + 1].isdigit()):
            j = i
            seen_dot = False
            while j < n and (src[j].isdigit() or (src[j] == "." and not seen_dot)):
                if src[j] == ".":
                    seen_dot = True
                j += 1
            # A bare unit letter immediately after digits makes it a duration.
            if not seen_dot and j < n and src[j] in DURATION_UNITS:
                is_unit_end = j + 1 >= n or not (src[j + 1].isalnum() or src[j + 1] == "_")
                if is_unit_end:
                    tokens.append(Token(TokType.DURATION, src[i:j + 1], i, j + 1))
                    i = j + 1
                    continue
            tokens.append(Token(TokType.NUMBER, src[i:j], i, j))
            i = j
            continue

        # Identifiers and keywords.
        if c.isalpha() or c == "_":
            j = i
            while j < n and (src[j].isalnum() or src[j] == "_"):
                j += 1
            word = src[i:j]
            kind = TokType.KEYWORD if word.lower() in KEYWORDS else TokType.IDENT
            tokens.append(Token(kind, word, i, j))
            i = j
            continue

        # Double-quoted identifiers, for column names that collide with keywords.
        if c == '"':
            j = src.find('"', i + 1)
            if j == -1:
                raise SqlError("Unterminated quoted identifier.", i, n)
            tokens.append(Token(TokType.IDENT, src[i + 1:j], i, j + 1))
            i = j + 1
            continue

        if c == "(":
            tokens.append(Token(TokType.LPAREN, c, i, i + 1))
            i += 1
            continue
        if c == ")":
            tokens.append(Token(TokType.RPAREN, c, i, i + 1))
            i += 1
            continue
        if c == ",":
            tokens.append(Token(TokType.COMMA, c, i, i + 1))
            i += 1
            continue

        matched = next((op for op in OPERATORS if src.startswith(op, i)), None)
        if matched:
            # '*' is its own token type so `SELECT *` and `a * b` can be told
            # apart by the parser without lookbehind hacks.
            ttype = TokType.STAR if matched == "*" else TokType.OP
            tokens.append(Token(ttype, matched, i, i + len(matched)))
            i += len(matched)
            continue

        raise SqlError(f"Unexpected character {c!r}.", i, i + 1)

    tokens.append(Token(TokType.EOF, "", n, n))
    return tokens


def parse_duration(text: str) -> int:
    """'5m' -> 300. Assumes the lexer already validated the shape."""
    return int(text[:-1]) * DURATION_UNITS[text[-1]]
